include/kafka/KafkaException.h

```diff
@@ -19,35 +21,55 @@ namespace KAFKA_API {
 class KafkaException: public std::exception
 {
 public:
-    KafkaException(const char* filename, int lineno, std::error_code ec, const std::string& errMsg)
-        : _errCode(ec)
-    {
-        _errMsg = Utility::getCurrentTime() + ": " + errMsg + " [" + std::to_string(ec.value())
-                  + "] (" + std::string(filename) + ":" + std::to_string(lineno) + ")";
-    }
+    KafkaException(const char* filename, std::size_t lineno, rd_kafka_resp_err_t respErr, const std::string& errMsg)
+        : _when(std::chrono::system_clock::now()),
+          _filename(filename),
+          _lineno(lineno),
+          _error(std::make_unique<SimpleError>(respErr, errMsg))
+    {}
+
+    KafkaException(const char* filename, int lineno, rd_kafka_resp_err_t respErr)
+        : KafkaException(filename, lineno, respErr, rd_kafka_err2str(respErr))
+    {}
 
-    KafkaException(const char* filename, int lineno, std::error_code ec)
-        : KafkaException(filename, lineno, ec, ec.message())
+    KafkaException(const char* filename, int lineno, const Error& error)
+        : _when(std::chrono::system_clock::now()),
+          _filename(filename),
+          _lineno(lineno),
+          _error(std::make_unique<Error>(error))
     {}
 
     /**
-     * Obtains the underlying error code.
+     * Obtains the underlying error.
      */
-    std::error_code error() const { return _errCode; }
+    const Error& error() const { return *_error; }
 
     /**
      * Obtains explanatory string.
      */
-    const char* what() const noexcept override { return _errMsg.c_str(); }
+    const char* what() const noexcept override
+    {
+        _what = Utility::getLocalTimeString(_when)  + ": " + _error->message() + " [" + std::to_string(_error->errorCode().value())
```

Comment:
Do we need to generate the error message every time we call what()?

Reply:
We have to choose between: (1) generate the error message every time the exception being constructed (2) generate the error message every time `what()` is called. Sounds option 1 is lazier. :-)

Reply:
The key is it's not likely we call `what()` twice after catching an exception, e.g. if we want to refer the error message multiple times, normally a string would be used to save the `what()`; in some scenarios, the `error()` would be fetched from the exception for error-handling, and `what()` would not even being called.